lib/Sema/CSBindings.cpp

```diff
@@ -714,12 +714,12 @@ ConstraintSystem::getPotentialBindingForRelationalConstraint(
   }
 
   // Make sure we aren't trying to equate type variables with different
-  // lvalue-binding rules.
+  // lvalue-binding rules unless we have no other options.
   if (auto otherTypeVar =
           type->lookThroughAllOptionalTypes()->getAs<TypeVariableType>()) {
```

Comment:
So I have tried just removing `->lookThroughAllOptionalTypes()` and running tests, looks like nothing is failing and even one of the diagnostics has been improved. I think that's what we want to do here.

Reply:
Hmm, I was seeing a failure with this snippet:
```
struct ImplicitMembers: Equatable {
    static var optional: ImplicitMembers? = ImplicitMembers()
}

func implicit(_ i: inout ImplicitMembers) {
    if i == .optional {}
}
```

Reply:
I'm not sure it should compile because the overload of `==` would be `(Wrapped?, Wrapped?) -> Bool` which means that base has to be `ImplicitMembers?`, right?

Reply:
Hm, it could pick the one on `Equatable` and the it would fail because of conversion constraint. I guess that check should only be done if the constraint is `Equal`...

Reply:
There's special logic to allow us to look through optionals at the base, so the code I posted compiles today. E.g., in CSApply, in `visitUnresolvedMemberExpr` we do this:

```
// If the member came by optional unwrapping, then unwrap the base type.
if (selected.choice.getKind() == OverloadChoiceKind::DeclViaUnwrappedOptional) {
  baseTy = baseTy->getOptionalObjectType();
  assert(baseTy && "got unwrapped optional decl from non-optional base?!");
}
```

Reply:
A simpler version of this behavior is just:
```
struct ImplicitMembers: Equatable {
    static var optional: ImplicitMembers? = ImplicitMembers()
}

let _: ImplicitMembers? = .optional
```

Reply:
I think it's a different problem because I don't think it's happening. That logic has to be added to be binding inference as well otherwise solver ends up failing this way when trying `(Wrapped?, Wrapped?) -> Bool` overload because it doesn't know that it's allowed to try `$T114` as well.

```
    (attempting type variable $T4 := $T114?
      (attempting fix [fix: define missing member named 'optional' based on its use] @ locator@0x7f801305be50 [UnresolvedMember@<stdin>:6:14 -> unresolved member])
```

Reply:
Looks like there is logic which prevents us from trying `$T114` there for `$T4` - https://github.com/apple/swift/blob/master/lib/Sema/CSBindings.cpp#L1131-L1134

Reply:
I don't think it has to match though because if we are trying to binding to something which is not-allowed to be l-value it should be okay even if type variable itself allows l-value.

Reply:
Not not vice versa.

Reply:
I think that condition has to become this:

```c++
         if (TypeVar->getImpl().canBindToLValue() ||
              !otherTypeVar->getImpl().canBindToLValue()) {
```

because if `TypeVar` allows l-value it doesn't matter what the type is, otherwise `otherTypeVar` can't allow l-value.

With that change your example type-checks with `->lookThroughAllOptionalTypes()` removed.

Reply:
Are you talking about the example I posted in the second comment? I'm still not able to compile that with that logic added and `lookThroughAllOptionalTypesRemoved()`...

Reply:
Yeah, that's the example I could compile with changes I mentioned (based on the trunk), here is the diff:

```
diff --git a/lib/Sema/CSBindings.cpp b/lib/Sema/CSBindings.cpp
index 46c02ad1b70..4797a4d5cd0 100644
--- a/lib/Sema/CSBindings.cpp
+++ b/lib/Sema/CSBindings.cpp
@@ -715,10 +715,9 @@ ConstraintSystem::getPotentialBindingForRelationalConstraint(

   // Make sure we aren't trying to equate type variables with different
   // lvalue-binding rules.
-  if (auto otherTypeVar =
-          type->lookThroughAllOptionalTypes()->getAs<TypeVariableType>()) {
+  if (auto otherTypeVar = type->lookThroughAllOptionalTypes()->getAs<TypeVariableType>()) {
     if (typeVar->getImpl().canBindToLValue() !=
-        otherTypeVar->getImpl().canBindToLValue())
+            otherTypeVar->getImpl().canBindToLValue())
       return None;
   }

@@ -1133,8 +1132,8 @@ bool TypeVarBindingProducer::computeNext() {
         // variable we will attempt to bind, both have the same
         // polarity with respect to being able to bind lvalues.
         if (auto otherTypeVar = objTy->getAs<TypeVariableType>()) {
-          if (TypeVar->getImpl().canBindToLValue() ==
-              otherTypeVar->getImpl().canBindToLValue()) {
+          if (TypeVar->getImpl().canBindToLValue() ||
+              !otherTypeVar->getImpl().canBindToLValue()) {
             addNewBinding(binding.withSameSource(objTy, binding.Kind));
           }
         } else {
```

Reply:
That diff still has the `lookThroughAllOptionalTypes()`, though, no?

Reply:
Ah, looks like I need to let this go for a bit, I have removed it and then re-added back in apparently...

Reply:
Ha no worries :) I'll continue looking at this a bit longer before I put it down for the night. Here's a reproducer that cuts out all the overload resolution cruft:

```
infix operator =%: ComparisonPrecedence

extension Optional {
    static func =%(lhs: Self, rhs: Self) -> Bool { return true }
}

struct ImplicitMembers {
    static var optional: ImplicitMembers? = ImplicitMembers()
}

func implicit(_ i: inout ImplicitMembers) {
    if i =% .optional {}
}
```

Reply:
Looks like this is indeed a different problem which was masked by the ordering in which bindings where picked before. There is a constraint `$T4 conv $T2` and `$T4` is going to be bound to `$T47?` which is correct but that means that solver would try to lookup `.optional` on `T47?` because it doesn't that that it can look through optional if it's unresolved member.

It looks like _currently_ `$T4` is only bound after generic parameter is bound to `ImplicitMembers`. It seems like `simplifyMemberConstraint` needs to handle situation when base is not yet completely resolved and delay simplification...

Reply:
I managed to get it to type check by adding this to perform member lookup:
```
  if (constraintKind == ConstraintKind::UnresolvedValueMember)
    if (instanceTy->getOptionalObjectType()->getAs<TypeVariableType>()) {
      MemberLookupResult result;
      result.OverallResult = MemberLookupResult::Unsolved;
      return result;
    }
```
Does that seem reasonable to you?

Reply:
Here is the change which allows this example to type-check correctly:

```
diff --git a/lib/Sema/CSBindings.cpp b/lib/Sema/CSBindings.cpp
index 46c02ad1b70..bb99b97afd9 100644
--- a/lib/Sema/CSBindings.cpp
+++ b/lib/Sema/CSBindings.cpp
@@ -715,8 +715,7 @@ ConstraintSystem::getPotentialBindingForRelationalConstraint(

   // Make sure we aren't trying to equate type variables with different
   // lvalue-binding rules.
-  if (auto otherTypeVar =
-          type->lookThroughAllOptionalTypes()->getAs<TypeVariableType>()) {
+  if (auto otherTypeVar = type->getAs<TypeVariableType>()) {
     if (typeVar->getImpl().canBindToLValue() !=
         otherTypeVar->getImpl().canBindToLValue())
       return None;
diff --git a/lib/Sema/CSSimplify.cpp b/lib/Sema/CSSimplify.cpp
index b85e02d5094..2d226867456 100644
--- a/lib/Sema/CSSimplify.cpp
+++ b/lib/Sema/CSSimplify.cpp
@@ -6590,6 +6590,12 @@ performMemberLookup(ConstraintKind constraintKind, DeclNameRef memberName,
       baseObjTy->is<AnyMetatypeType>() &&
       constraintKind == ConstraintKind::UnresolvedValueMember) {
     if (auto objectType = instanceTy->getOptionalObjectType()) {
+      if (objectType->isTypeVariableOrMember()) {
+        MemberLookupResult result;
+        result.OverallResult = MemberLookupResult::Unsolved;
+        return result;
+      }
+
       if (objectType->mayHaveMembers()) {
         LookupResult &optionalLookup = lookupMember(objectType, memberName);
         for (auto result : optionalLookup)
```

Reply:
Thanks @xedin, that looks totally reasonable and can confirm that it works! I've updated the PR here, and added the reproducer to the tests.

Reply:
Thanks for looking into this!

Reply:
Thank you for puzzling through it with me! Now to see if it has any impact on implicit member chains... 🤞